flash_io.c

```diff
@@ -0,0 +1,622 @@
+
+/*
+ *   The main purpose of saving user data to flash memory is to
+ *   handle unexpected power loss.
+ *   In such cases, the device can reboot with previous config.
+ *
+ *   To read from flash memory, TAG must be a valid entry, 
+ *   and target must match the current setup.
+ *
+ *   Additionally, we only read from flash if  the last saved record
+ *   has the same configuration as the current running config.
+ *   This prevents from booting with outdated data.
+ *
+ *  Memory layout:
+ *
+ *    TAG                    -> Distinguishes between random bytes and a valid entry
+ *    targets                -> Indicates which targets are initialized
+ *
+ *                              TAG and targets are consts, set at compile-time.
+ *                              The remaining fields are copied values from the runtime
+ *
+ *    log_bits               -> These three fields are always written to memory.
+ *    pwm_level              -> They describe the device driver state, and are thus
+ *    ceiling_pwm            -> always written in.
+ *
+ *                           -> Below this line, every fields are written only
+ *                              if corrensponding driver is activated
+ *                              (e.g., pwm_water is saved only if
+ *                               CONFIG_WATER is set)
+ *
+ *    pwm_water
+ *
+ *    magnetron_pulse_count
+ *    magnetron_deadline
+ *
+ *    pilot_is_enabled
+ *    pilot_des_temp
+ *
+ *    mapper_is_enabled
+ *    mapper_max_pwm_temp
+ *
+ *  End of memory layout
+ *
+ *    Data is written as a continuous array with no padding between fields.
+ *    So, if CONFIG_WATER and CONFIG_AUTO_MAPPER are enabled,
+ *    the memory layout will look like:
+ *
+ *      TAG
+ *      targets
+ *      log_bits
+ *      pwm_level
+ *      ceiling_pwm
+ *      pwm_water
+ *      mapper_is_enabled
+ *      mapper_max_pwm_temp
+ *
+ *
+ *
+ *    Flash memory limitations:
+ *
+ *     Flash memory cannot be rewritten directly.
+ *     Once data is written, it cannot be overwritten
+ *     without erasing the entire flash sector.
+ *     This is a property of NOR/NAND flash, where
+ *     each bit is initialized to 1 by default.
+ *
+ *     The pico-sdk requires that the write address is aligned to a
+ *     256-byte boundary (flash page), and the data must be written in
+ *     multiples of 256 bytes.
+ *
+ *     Additionally, the pico-sdk requires that erasures start at a
+ *     4096-byte boundary (flash sector) and erase a multiple of 4096 bytes.
+ *
+ *    (ref: https://www.raspberrypi.com/documentation/pico-sdk/hardware.html#group_hardware_flash_1ga5b88461b126b1c226121146c87d5abe2 )
+ *
+ *    Because of this constraints, this driver uses 2 flash sectors.
+ *    (configured in memmap.ld)
+ *
+ *
+ *    Algorithm overview:
+ *
+ *     The algorithm writes to memory whenever the runtime
+ *     configuration changes.
+ *     Changes are checked at intervals defined by FLASH_WRITE_MS.
```

Comment:
Can't we update the flash when we update the fields that are saved and only if the value changed?

Reply:
We only write to flash memory if the value has changed, and we check for changes at intervals of FLASH_WRITE_MS to avoid overusing the flash memory. Often, during runtime, users may change multiple options at once (e.g., temp = 850, auto = 1), so we want to batch these updates and write them all at once. This prevents multiple writes to flash. For example, if a user changes auto from 1 to 0 and then back to 1 before the next flash check, the flash won't be updated unnecessarily.